src/cli/commands/build-static.ts

```
@@ -123,17 +126,54 @@ type KVStoreItemDesc = {
-      console.log(`✔️ Submitted ${text ? 'text' : 'binary'} asset "${staticFilePath}" to KV Store with key "${kvStoreKey}".`)
+
+  const maxConcurrent = 12;
+  let index = 0;
+
+  async function worker() {
+    while (index < kvStoreItems.length) {
+      const currentIndex = index;
+      index = index + 1;
+      const { kvStoreKey, staticFilePath, text } = kvStoreItems[currentIndex];
+
+      try {
+        await attemptWithRetries(
+          async() => {
+            if (await kvStoreEntryExists(fastlyApiContext, kvStoreName, kvStoreKey)) {
+              console.log(`✔️ Asset already exists in KV Store with key "${kvStoreKey}".`);
+              return;
+            }
+            const fileData = fs.readFileSync(staticFilePath);
+            await kvStoreSubmitFile(fastlyApiContext, kvStoreName, kvStoreKey, fileData);
+            console.log(`✔️ Submitted ${text ? 'text' : 'binary'} asset "${staticFilePath}" to KV Store with key "${kvStoreKey}".`)
+          },
+          {
+            onAttempt(attempt) {
+              if (attempt > 0) {
+                console.log(`Attempt ${attempt + 1} for: ${kvStoreKey}`);
+              }
+            },
+            onRetry(attempt, err, delay) {
+              let statusMessage = 'unknown';
+              if (err instanceof FetchError) {
+                statusMessage = `HTTP ${err.status}`;
+              } else if (err instanceof TypeError) {
+                statusMessage = 'transport';
+              }
+              console.log(`Attempt ${attempt + 1} for ${kvStoreKey} gave retryable error (${statusMessage}), delaying ${delay} ms`);
+            },
+          }
+        );
+      } catch (err) {
+        const e = err instanceof Error ? err : new Error(String(err));
+        console.error(`❌ Failed: ${kvStoreKey} → ${e.message}`);
+        console.error(e.stack);
+      }
     }
   }
+
+  const workers = Array.from({ length: maxConcurrent }, () => worker());
```

Since `kvStoreItems` is not sharded for use by the workers, each worker is going to attempt to upload the same items. Initially, wouldn't that mean that all of them will attempt to upload the first item in parallel?

It might be simplest to tell each worker what its worker number is, and then have it use that as an initial index value for `kvStore.items`, and all workers would use `maxConcurrent` as a step value for the index.

---

Even though the workers are async and run in parallel over time, the important part that grabs the next index happens synchronously, one at a time. JavaScript's event loop guarantees that `index = index + 1` executes without interruption. So each worker gets a unique item, and no items are duplicated.

However, I can add a comment like this to make it less scary to readers of the code:
```
// This loop is safe because JavaScript is single-threaded.
// Workers pull work one item at a time using a shared index.
```

Although your suggestion is a valid way of splitting the work, in JavaScript’s async model, using a shared counter like `index = index + 1` is safe and gives us dynamic work balancing, which can help if some requests take longer than others. It also simplifies retries and avoids leaving gaps in the work if a task fails early.

---

Ahh, thank you for the education! I had missed the fact that `index` is declared outside the `worker` function, which means it is shared by all instances of that function (and thus access to it is serialized, as you explained).

I think adding a comment on the declaration of `index` itself would be sufficient, something like "This variable is declared the module level so that all instances of the 'worker' function will share it."